SciMLLogging Integration

jClugstor opened the pull request:
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API
  
## Additional context

```julia 
using LinearSolve, BenchmarkTools
using SciMLLogging: Verbosity
n = 4
A = rand(n, n)
b1 = rand(n);
b2 = rand(n);
prob = LinearProblem(A, b1)
```

With this PR:
```julia
@btime solve(prob, verbose=false)
365.015 ns (11 allocations: 928 bytes)

using SciMLLogging: Verbosity
@btime solve(prob, verbose = LinearVerbosity(Verbosity.None()))
363.893 ns (12 allocations: 960 bytes)

@btime solve(prob, verbose=true)
 432.445 ns (15 allocations: 1.03 KiB)

@report_opt solve(prob, verbose = LinearVerbosity(Verbosity.None()))
No errors detected

@report_opt solve(prob, verbose = false)
No errors detected

@profview for i in 1:10000000 init(prob, verbose = false) end 
```
<img width="774" height="151" alt="NewPRProfile" src="https://github.com/user-attachments/assets/49426d93-b5aa-4317-b908-e93f30f1c19f" />

From the profile we can see that the constructor for LinearVerbosity doesn't even get called from `__init`, which should indicate that it got compiled out. 

With current main (Bool verbosity system)

```julia
`@btime solve(prob, verbose=false)`
 `359.857 ns (11 allocations: 928 bytes)`

`@btime solve(prob, verbose=true)`
` 329.573 ns (11 allocations: 928 bytes)`

@profview for i in 1:10000000 init(prob, verbose = false) end 
```
<img width="772" height="154" alt="OldVerbosityProfile" src="https://github.com/user-attachments/assets/a245cf82-2d9c-486b-9d5d-61ff5b5a90cd" />

jClugstor commented:
I rewrote SciMLLogging to get rid of the Moshi.jl dependency, and to make it easier to implement Verbosity "presets". I also rewrote the stuff in LinearVerbosity. This way has fewer dependencies, and is still able to compile out the logging code. Better yet, even if logging is enabled, if the code goes through an `@SciMLMessage` where the verbosity toggle is set to `Verbosity.Silent()` the it doesn't go through any of the SciMLLogging code because it gets compiled away, at least in the testing I've done so far. Which means that no time is wasted getting the verbosity type, checking if it's None, etc. anymore. 

With this PR:
```julia 
n = 4
A = rand(n, n)
b1 = rand(n);
b2 = rand(n);
prob = LinearProblem(A, b1)

@b solve(prob, verbose = false)
373.308 ns (12 allocs: 1008 bytes)

@b solve(prob, verbose = true )
364.000 ns (12 allocs: 1008 bytes)

@b solve(prob, verbose = LinearVerbosity()) 
378.214 ns (13 allocs: 1.000 KiB)

@b solve(prob)
378.121 ns (12 allocs: 1008 bytes)

verb = LinearVerbosity()

@b solve(prob, verbose = verb)
527.854 ns (13 allocs: 960 bytes)

opt = @report_opt solve(prob, LUFactorization(), verbose = LinearVerbosity())
No errors detected

opt = @report_opt solve(prob, LUFactorization(), verbose = LinearVerbosity(Verbosity.None()))
No errors detected

opt = @report_opt solve(prob, LUFactorization(), verbose = false)
No errors detected

cache = init(prob, LUFactorization())

@b solve!(cache) 
101.288 ns (1 allocs: 48 bytes)

A = [1.0 0 0 0
    0 1 0 0
    0 0 1 0
    0 0 0 0]
b = rand(4)
prob = LinearProblem(A, b)

verb = LinearVerbosity(default_lu_fallback = Verbosity.Silent())

cache = init(prob, verbose = verb)

@b solve!(cache)
104.880 ns (1 allocs: 48 bytes)

cache = init(prob, verbose=false)

@b solve!(cache)
104.812 ns (1 allocs: 48 bytes)
```

LinearSolve v3.40.1:
```julia
n = 4
A = rand(n, n)
b1 = rand(n);
b2 = rand(n);
prob = LinearProblem(A, b1)

@b solve(prob, verbose = false)
252.902 ns (11 allocs: 928 bytes)

@b solve(prob, verbose = true )
250.919 ns (11 allocs: 928 bytes)

A = [1.0 0 0 0
    0 1 0 0
    0 0 1 0
    0 0 0 0]
b = rand(4)
prob = LinearProblem(A, b)

cache = init(prob, verbose = false)

104.171 ns (1 allocs: 48 bytes)
```

So the creation of the LinearVerbosity adds a little bit of overhead to `init`, but once the cache is created, if there's no printing, there's no overhead.

jClugstor commented:
I should just get rid of that testset?

jClugstor commented:
Ah yeah, I have it check if the extension is loaded and don't run the tests if it isn't. 
Ok, I'll add those in as well.

ChrisRackauckas commented:
> # Standard preset - warnings and errors (should see LU fallback warning)

LU fallback warning should be detailed and above only. It's non-critical and most users shouldn't handle it, it's just a behind the scenes thing. It's even one of the examples in https://docs.sciml.ai/SciMLLogging/dev/manual/presets/#Preset-Level-Examples

jClugstor commented:
Ah, missed that. Fixed now

jClugstor commented:
I've added three new verbosity options
- convergence_failure - Messages when iterative solvers fail to converge
- max_iters - Messages when iterative solvers reach maximum iterations
- solver_failure - Messages when solvers fail for reasons other than convergence

All three:
Are in the numerical group
Default to `WarnLevel()` in Standard, Detailed, and All presets
Are `Silent()` in Minimal and None presets

I added a `solver_failure` message whenever the retcode is set to `Failure`, `convergence_failure` for `ConvergenceFailure`, and `max_iters` for `MaxIters`.

ChrisRackauckas commented on src/common.jl:
```diff
@@ -324,6 +324,20 @@ function __init(prob::LinearProblem, alg::SciMLLinearSolveAlgorithm,
 
+    if verbose isa Bool
+        #@warn "Using `true` or `false` for `verbose` is being deprecated. Please use a `LinearVerbosity` type to specify verbosity settings.
+        # For details see the verbosity section of the common solver options documentation page."
```
warn here

jClugstor commented:
Integration tests are passing now, and it looks like the rest of the failures are from mooncake/enzyme on 1.12
